packages/hash/frontend/src/pages/verification.page.tsx

```diff
@@ -0,0 +1,186 @@
+import { useRouter } from "next/router";
+import { useEffect, FormEventHandler, useState, useMemo } from "react";
+import {
+  UpdateVerificationFlowWithCodeMethodBody,
+  VerificationFlow,
+} from "@ory/client";
+import { isUiNodeInputAttributes } from "@ory/integrations/ui";
+import { Typography, Container, Box } from "@mui/material";
+import { TextField } from "@hashintel/hash-design-system";
+
+import { getPlainLayout, NextPageWithLayout } from "../shared/layout";
+import {
+  createFlowErrorHandler,
+  gatherUiNodeValuesFromFlow,
+  oryKratosClient,
+} from "./shared/ory-kratos";
+import { Button } from "../shared/ui";
+import { useLogoutFlow } from "../components/hooks/useLogoutFlow";
+
+const VerificationPage: NextPageWithLayout = () => {
+  // Get ?flow=... from the URL
+  const router = useRouter();
+
+  const {
+    return_to: returnTo,
+    flow: flowId,
+    // Refresh means we want to refresh the session. This is needed, for example, when we want to update the password
+    // of a user.
+    refresh,
+    // AAL = Authorization Assurance Level. This implies that we want to upgrade the AAL, meaning that we want
+    // to perform two-factor authentication/verification.
+    aal,
+  } = router.query;
+
+  const [flow, setFlow] = useState<VerificationFlow>();
+  const [code, setCode] = useState<string>();
+  const [errorMessage, setErrorMessage] = useState<string>();
+
+  const handleFlowError = useMemo(
+    () =>
+      createFlowErrorHandler({
+        router,
+        flowType: "verification",
+        setFlow,
+        setErrorMessage,
+      }),
+    [router, setFlow, setErrorMessage],
+  );
+
+  // This might be confusing, but we want to show the user an option
+  // to sign out if they are performing two-factor authentication!
+  const { logout } = useLogoutFlow([aal, refresh]);
+
+  const extractFlowCodeValue = (flowToSearch: VerificationFlow | undefined) => {
+    const uiCode = flowToSearch?.ui?.nodes.find(
+      ({ attributes }) =>
+        isUiNodeInputAttributes(attributes) && attributes.name === "code",
+    );
+    if (uiCode?.attributes && "value" in uiCode.attributes) {
+      setCode(String(uiCode.attributes.value));
+    }
+  };
+
+  useEffect(() => {
+    // If the router is not ready yet, or we already have a flow, do nothing.
+    if (!router.isReady || flow) {
+      return;
+    }
+
+    // If ?flow=.. was in the URL, we fetch it
+    if (flowId) {
+      oryKratosClient
+        .getVerificationFlow({ id: String(flowId) })
+        .then(({ data }) => {
+          setFlow(data);
+          extractFlowCodeValue(data);
+        })
+        .catch(handleFlowError);
+      return;
+    }
+
+    // Otherwise we initialize it
+    oryKratosClient
+      .createBrowserVerificationFlow({
+        returnTo: returnTo ? String(returnTo) : undefined,
+      })
+      .then(({ data }) => {
+        setFlow(data);
+        extractFlowCodeValue(data);
```

Comment:
Could we programmatically submit the form when it is populated from the query params for the first time? This would prevent the user from having to click anything.

Reply:
I think we could, yes. But the reason to use these codes in the first place is to prevent mail apps to auto-open the links on the user's behalf. I appreciate that we'd gate this behind the `/login` interface and not submit, but I think we're better off not automating it for now